File: mySPOJ/CiekawaWyliczanka.cpp
///////////////////////////////////////////////////////
/// NOTES: ////////////////////////////////////////////
/// TASK ID = 751 , NAME: CIEKAWA WYLICZANKA //////////
/// link: https://pl.spoj.com/problems/ETI06F2 ////////
/// TIME RESULT FOR THIS SOLUTION = 0,00s /////////////
///////////////////////////////////////////////////////

#include <iostream>
#include <string>
#include <bitset>
#include <boost/algorithm/string.hpp>
using namespace std;

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);

	int n, i, baseNr, length, diff, baseLen, diffLen;
	string binaryN, diffChars, baseChars = "";
	bool ifContains;

	cin >> n;
	binaryN = bitset<32>(n).to_string();
	while (binaryN[0] == '0') binaryN.erase(0, 1);
	//cout << "binaryN=" << binaryN << "\n";
	length = binaryN.length();
	//cout << "length=" << length << "\n";

	ifContains = boost::algorithm::contains(binaryN, "0");
	if (ifContains) {
		length--;
		for (i = 0; i < length; i++) baseChars.append("5");
		//cout << "baseChars=" << baseChars << "\n";
		baseNr = 1;
		for (i = 0; i < length; i++) baseNr = baseNr * 2;
		baseNr--;
		//cout << "baseNr=" << baseNr << "\n";
		diff = n - baseNr;
		//cout << "diff=" << diff << "\n";
		baseLen = baseChars.length();
		//cout << "baseLen=" << baseLen << "\n";
		diffChars = bitset<32>(diff).to_string();
		while (diffChars[0] == '0') diffChars.erase(0, 1);
		//cout << "diffChars=" << diffChars << "\n";
		diffLen = diffChars.length();
		//cout << "diffLen=" << diffLen << "\n";
		for (int i = 0; i < diffLen; i++)
			baseChars[baseLen - i - 1] = (char)(baseChars[baseLen - i - 1]
				+ diffChars[diffLen - i - 1] - 48);

		cout << baseChars;
	}
	else
		for (i = 0; i < length; i++) cout << (5);

	return 0;
}

File: src/main/java/spoj/CiekawaWyliczanka.java
///////////////////////////////////////////////////////
/// NOTES: ////////////////////////////////////////////
/// TASK ID = 751 /////////////////////////////////////
/// NAME: CIEKAWA WYLICZANKA //////////////////////////
/// link: https://pl.spoj.com/problems/ETI06F2 ////////
/// TIME RESULT FOR THIS SOLUTION = 0,91s /////////////
///////////////////////////////////////////////////////
package spoj;

import java.io.BufferedOutputStream;
import java.io.DataInputStream;
import java.io.IOException;


public class CiekawaWyliczanka {

    static class Reader {
        final private int BUFFER_SIZE = 1 << 16;
        private DataInputStream din;
        private byte[] buffer;
        private int bufferPointer, bytesRead;

        public Reader() {
            din = new DataInputStream(System.in);
            buffer = new byte[BUFFER_SIZE];
            bufferPointer = bytesRead = 0;
        }

        public int nextInt() throws IOException {
            int num = 0;
            byte c = read();
            while (c <= ' ') c = read();
            do {
                num = num * 10 + c - '0';
            } while ((c = read()) >= '0' && c <= '9');
            return num;
        }

        private void fillBuffer() throws IOException {
            bytesRead = din.read(buffer, bufferPointer = 0, BUFFER_SIZE);
            if (bytesRead == -1) buffer[0] = -1;
        }

        private byte read() throws IOException {
            if (bufferPointer == bytesRead) fillBuffer();
            return buffer[bufferPointer++];
        }
    }

    public static void main(String[] args) throws IOException {
        Reader rd = new Reader();
        StringBuilder strb = new StringBuilder();
        BufferedOutputStream out = new BufferedOutputStream(System.out);

        int n = rd.nextInt();

        String binaryN = Integer.toBinaryString(n);
        int length = binaryN.length();
        if (binaryN.contains("0")) {
            length--;
            for (int i = 0; i < length; i++) strb.append(5);
            int baseNr = 1;
            for (int i = 0; i < length; i++) baseNr = baseNr * 2;
            baseNr--;
            int diff = n - baseNr;
            char[] baseChars = strb.toString().toCharArray();
            char[] diffChars = Integer.toBinaryString(diff).toCharArray();
            int baseLen = baseChars.length;
            int diffLen = diffChars.length;
            for (int i = 0; i < diffLen; i++)
                baseChars[baseLen - i - 1] = (char) (baseChars[baseLen - i - 1] + diffChars[diffLen - i - 1] - 48);
            strb.delete(0,strb.length());
            strb.append(baseChars);
        } else
            for (int i = 0; i < length; i++) strb.append(5);

        out.write(strb.toString().getBytes());
        out.flush();
    }
}
